### PR TITLE
[SC-675] Add Aave support

### PR DESCRIPTION
**Pull request by hexonaut**
Adding atoken support and integration tests for Aave main market.

### CODE REVIEW

File: src/MainnetController.sol [hunk 1]
```diff
@@ -253,6 +261,51 @@ contract MainnetController is AccessControl {
...
+
+        // Withdraw underlying from Aave pool, decode resulting amount withdrawn.
+        // Assumes proxy has adequate atokens.
+        amountWithdrawn = abi.decode(
+            proxy.doCall(
+                address(pool),
+                abi.encodeCall(pool.withdraw, (IAToken(atoken).UNDERLYING_ASSET_ADDRESS(), amount, address(proxy)))
+            ),
+            (uint256)
+        );
```
Comment (lucas-manuel) [inline src/MainnetController.sol:306]: Small formatting change to get under line limit
```suggestion
        amountWithdrawn = abi.decode(
            proxy.doCall(
                address(pool),
                abi.encodeCall(
                    pool.withdraw,
                    (IAToken(atoken).UNDERLYING_ASSET_ADDRESS(), amount, address(proxy))
                )
            ),
            (uint256)
        );
```

File: src/MainnetController.sol [hunk 2]
```diff
@@ -253,6 +261,51 @@ contract MainnetController is AccessControl {
         );
     }
 
+    /**********************************************************************************************/
+    /*** Relayer Aave functions                                                                 ***/
+    /**********************************************************************************************/
+
+    function depositAave(address atoken, uint256 amount)
```
Comment (lucas-manuel) [inline src/MainnetController.sol:268]: `aToken`

File: src/MainnetController.sol [hunk 3]
```diff
@@ -253,6 +261,51 @@ contract MainnetController is AccessControl {
...
+        onlyRole(RELAYER)
+        isActive
+        rateLimited(
+            RateLimitHelpers.makeAssetKey(LIMIT_AAVE_DEPOSIT, atoken),
+            amount
+        )
+    {
+        IERC20 underlying = IERC20(IAToken(atoken).UNDERLYING_ASSET_ADDRESS());
+        IAavePool pool    = IAavePool(IATokenWithPool(atoken).POOL());
+
```
Comment (lucas-manuel) [inline src/MainnetController.sol:279]:
```solidity
        IERC20    underlying = IERC20(IAToken(atoken).UNDERLYING_ASSET_ADDRESS());
        IAavePool pool       = IAavePool(IATokenWithPool(atoken).POOL());
```

File: src/MainnetController.sol [hunk 4]
```diff
@@ -36,6 +39,10 @@ interface IPSMLike {
     function to18ConversionFactor() external view returns (uint256);
 }
 
+interface IATokenWithPool {
```
Comment (lucas-manuel) [inline src/MainnetController.sol:42]: Can add is `IAToken` and use for both

File: src/MainnetController.sol [hunk 5]
```diff
@@ -253,6 +261,51 @@ contract MainnetController is AccessControl {
...
+        IERC20 underlying = IERC20(IAToken(atoken).UNDERLYING_ASSET_ADDRESS());
+        IAavePool pool    = IAavePool(IATokenWithPool(atoken).POOL());
+
+        // Approve underlying to Aave pool from the proxy (assumes the proxy has enough underlying).
+        proxy.doCall(
+            address(underlying),
+            abi.encodeCall(underlying.approve, (address(pool), amount))
+        );
+
+        // Deposit underlying into Aave pool, proxy receives atokens
```
Comment (lucas-manuel) [inline src/MainnetController.sol:286]: aTokens

File: test/mainnet-fork/Aave.t.sol [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+}
+
+contract AaveV3MainMarketDepositSuccessTests is AaveV3MainMarketBaseTest {
+
+    function test_aave_usds_deposit() public {
+        deal(Ethereum.USDS, address(almProxy), 1_000_000e18);
+
+        assertEq(ausds.balanceOf(address(almProxy)),                       0);
+        assertEq(IERC20(Ethereum.USDS).balanceOf(address(almProxy)),       1_000_000e18);
```
Comment (lucas-manuel) [inline test/mainnet-fork/Aave.t.sol:94]: Assert atoken balances of underlying everywhere

File: test/mainnet-fork/Aave.t.sol [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        assertEq(ausds.balanceOf(address(almProxy)),                 600_000e18);
+        assertEq(IERC20(Ethereum.USDS).balanceOf(address(almProxy)), 400_000e18);
+
+        // Withdraw all
+        vm.prank(relayer);
+        assertEq(mainnetController.withdrawAave(ATOKEN_USDS, type(uint256).max), 600_000e18);
+
+        assertEq(ausds.balanceOf(address(almProxy)),                 0);
+        assertEq(IERC20(Ethereum.USDS).balanceOf(address(almProxy)), 1_000_000e18);
+    }
```
Comment (lucas-manuel) [inline test/mainnet-fork/Aave.t.sol:167]: Lets warp in these tests to get a more realistic sense of the rebasing tokens, also on top of this adding aToken balance of underlying as per comment above
```suggestion
    function test_aave_usds_withdraw() public {
        deal(Ethereum.USDS, address(almProxy), 1_000_000e18);
        vm.prank(relayer);
        mainnetController.depositAave(ATOKEN_USDS, 1_000_000e18);

        skip(1 days);

        uint256 fullBalance = ausds.balanceOf(address(almProxy));

        assertGe(fullBalance, 1_000_000e18);

        assertEq(ausds.balanceOf(address(almProxy)),                 fullBalance);
        assertEq(IERC20(Ethereum.USDS).balanceOf(address(almProxy)), 0);

        // Partial withdraw
        vm.prank(relayer);
        assertEq(mainnetController.withdrawAave(ATOKEN_USDS, 400_000e18), 400_000e18);

        assertEq(ausds.balanceOf(address(almProxy)),                 fullBalance - 400_000e18);
        assertEq(IERC20(Ethereum.USDS).balanceOf(address(almProxy)), 400_000e18);

        // Withdraw all
        vm.prank(relayer);
        assertEq(mainnetController.withdrawAave(ATOKEN_USDS, type(uint256).max), fullBalance - 400_000e18);

        assertEq(ausds.balanceOf(address(almProxy)),                 0);
        assertEq(IERC20(Ethereum.USDS).balanceOf(address(almProxy)), fullBalance);
    }
```

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: src/MainnetController.sol (hunk 1 above)
Comment (hexonaut) [reply src/MainnetController.sol:306]: changed

File: src/MainnetController.sol (hunk 2 above)
Comment (hexonaut) [reply src/MainnetController.sol:268]: I would rather specify this as Aave to be clear at a high level.

File: src/MainnetController.sol (hunk 3 above)
Comment (hexonaut) [reply src/MainnetController.sol:279]: changed

File: src/MainnetController.sol (hunk 4 above)
Comment (hexonaut) [reply src/MainnetController.sol:42]: done

File: test/mainnet-fork/Aave.t.sol (hunk 6 above)
Comment (hexonaut) [reply test/mainnet-fork/Aave.t.sol:94]: To me this is a technical detail of the protocol and unimportant at this level. The interface is "you send me tokens and I send back receipt aTokens". In a similar way we don't care where the tokens end up for ERC4626 vaults we shouldn't care where they end up in Aave-style as we trust the protocol has implemented the relationship correctly.

File: test/mainnet-fork/Aave.t.sol (hunk 7 above)
Comment (hexonaut) [reply test/mainnet-fork/Aave.t.sol:167]: Similarly I'm not sure how much value this provides. We need to measure that we deposit and get a receipt token, but beyond that it's not really important how the underlying protocol works.

**Review by lucas-manuel — APPROVED** [review]

**Pull request merged**